src/IO/StackDataWriter.cs

```diff
@@ -0,0 +1,451 @@
+
+            Position += 4;
+        }
+
+        [MethodImpl(IMPL_OPTION)]
+        public void WriteUnicodeLE(string str)
+        {
+            EnsureSize((str.Length + 1) * 2);
+
+            for (int i = 0; i < str.Length; ++i)
+            {
+                char c = str[i];
+
+                if (c != '\0')
+                {
+                    WriteUInt16LE(c);
+                }
+            }
+
+            WriteUInt16LE(0x00);
+        }
+
+        [MethodImpl(IMPL_OPTION)]
+        public void WriteUnicodeLE(string str, int length)
+        {
+            EnsureSize(length);
+
+            for (int i = 0; i < length && i < str.Length; ++i)
+            {
+                char c = str[i];
+
+                if (c != '\0')
+                {
+                    WriteUInt16LE(c);
+                }
+            }
+
+            for (int i = str.Length; i < length; ++i)
+            {
+                WriteUInt16LE(0x00);
+            }
+        }
+
+
+
+
+        /* Big Endian */
+
+        [MethodImpl(IMPL_OPTION)]
+        public void WriteUInt16BE(ushort b)
+        {
+            EnsureSize(2);
+
+            _buffer[Position] = (byte)(b >> 8);
+            _buffer[Position + 1] = (byte)b;
+
+            Position += 2;
+        }
+
+        [MethodImpl(IMPL_OPTION)]
+        public void WriteInt16BE(short b)
+        {
+            EnsureSize(2);
+
+            _buffer[Position] = (byte)(b >> 8);
+            _buffer[Position + 1] = (byte)b;
+
+            Position += 2;
+        }
+
+        [MethodImpl(IMPL_OPTION)]
+        public void WriteUInt32BE(uint b)
+        {
+            EnsureSize(4);
+
+            _buffer[Position] = (byte)(b >> 24);
+            _buffer[Position + 1] = (byte)(b >> 16);
+            _buffer[Position + 2] = (byte)(b >> 8);
+            _buffer[Position + 3] = (byte)b;
+
+            Position += 4;
+        }
+
+        [MethodImpl(IMPL_OPTION)]
+        public void WriteInt32BE(int b)
+        {
+            EnsureSize(4);
+
+            _buffer[Position] = (byte)(b >> 24);
+            _buffer[Position + 1] = (byte)(b >> 16);
+            _buffer[Position + 2] = (byte)(b >> 8);
+            _buffer[Position + 3] = (byte)b;
+
+            Position += 4;
+        }
+
+        [MethodImpl(IMPL_OPTION)]
+        public void WriteUnicodeBE(string str)
+        {
+            EnsureSize((str.Length + 1) * 2);
+
+            for (int i = 0; i < str.Length; ++i)
+            {
+                char c = str[i];
+
+                if (c != '\0')
+                {
+                    WriteUInt16BE(c);
+                }
+            }
+
+            WriteUInt16BE(0x00);
+        }
+
+        [MethodImpl(IMPL_OPTION)]
+        public void WriteUnicodeBE(string str, int length)
+        {
+            EnsureSize(length);
+
+            for (int i = 0; i < length && i < str.Length; ++i)
+            {
+                char c = str[i];
+
+                if (c != '\0')
```

Comment:
When I did testing on this, I found that doing an `indexOf` from a `ReadOnlySpan<char>`, slicing from that point, counting the chars and multiplying by 2, then getting an equal sized buffer and just doing `Encoding.Unicode.GetBytes()` to write directly to it was actually faster.

I highly recommend benchmarks for this kind of thing since regular for loops are slower than whatever magic they are doing under the hood (from my experience). This is also specific to .NET Core/5, I am not sure how well .NET Framework performs for encoding.

Reply:
https://github.com/andreakarasho/ClassicUO/pull/1370/commits/9ea4049a7c4d4b4c079b305267694c55a001bb88
Done some mod, what do you think about?

Reply:
Looks great!